File: BackEnd/Malaz/app/Models/Booking.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Carbon;

class Booking extends Model
{
    protected $fillable = [
        'user_id',
        'property_id',
        'check_in',
        'check_out',
        'status',
        'total_price',
    ];

    protected $casts = [
        'check_in' => 'date',
        'check_out' => 'date',
    ];

    // A booking belongs to a user
    public function user()
    {
        return $this->belongsTo(User::class);
    }

    // A booking belongs to a property
    public function property()
    {
        return $this->belongsTo(Property::class);
    }

    /**
     * Auto-update the booking status based on dates
     * Call this method whenever you need to ensure status is current
     */
    public function autoUpdateStatus()
    {
        $today = Carbon::today();
        $checkIn = Carbon::parse($this->check_in);
        $checkOut = Carbon::parse($this->check_out);

        // Update confirmed → ongoing
        if ($this->status === 'confirmed' && $today->between($checkIn, $checkOut)) {
            $this->status = 'ongoing';
            $this->save();
        }
        // Update ongoing → completed
        elseif ($this->status === 'ongoing' && $today->greaterThan($checkOut)) {
            $this->status = 'completed';
            $this->save();
        }

        return $this;
    }
}

File: frontend/malaz/lib/domain/entities/booking/booking.dart
import 'package:equatable/equatable.dart';
import 'package:malaz/domain/entities/user/user_entity.dart';
import '../apartment/apartment.dart';

class Booking extends Equatable {
  final int? id;
  final int? propertyId;
  final DateTime? checkIn;
  final DateTime? checkOut;
  final num? price;
  final int? userId;
  final UserEntity? user;
  final String? status;
  final String? totalPrice;
  final Apartment? apartment;

  const Booking({
    this.id,
    this.propertyId,
    this.checkIn,
    this.checkOut,
    this.price,
    this.userId,
    this.user,
    this.status,
    this.totalPrice,
    this.apartment,
  });

  @override
  List<Object?> get props => [
    id,
    propertyId,
    checkIn,
    checkOut,
    price,
    userId,
    user,
    status,
    totalPrice,
    apartment,
  ];
}